app/_src/kubernetes-ingress-controller/guides/security/verify-upstream-tls.md

```diff
@@ -0,0 +1,392 @@
+
+{% include /md/kic/ca-certificates-note.md %}
+
+```shell
+echo '
+apiVersion: v1
+kind: Secret
+metadata:
+  name: root-ca
+  labels:
+    konghq.com/ca-cert: "true" # This label is required for the CA certificate to be recognized by Kong
+  annotations:
+    kubernetes.io/ingress.class: kong
+data:
+  cert: '$(base64 -w0 ./certs/root.crt)'
+  # An arbitrary ID for the certificate 
+  id: '$(printf "bf6e0f14-78cd-45ad-9325-87ec7ef7b890" | base64 -w0)'
+' | kubectl apply -f -
+```
+
+The results should look like this.
+
+```text
+secret/root-ca created
+```
+
+Now, associate the root CA certificate with the service passing its name to `konghq.com/ca-certificates` annotation.
+
+{:.note}
+> The `konghq.com/ca-certificates` annotation is a comma-separated list of CA certificate names. You can add multiple
+> CA certificates to the list.
+
+```shell
+kubectl annotate service echo konghq.com/ca-certificates='root-ca'
+```
+
+The results should look like this.
+
+```text
+service/echo annotated
+```
+
+Now, when you issue the same request as before, you should see a successful response.
+
+```shell
+curl -H Host:kong.example $PROXY_IP/echo
+```
+
+```text
+Welcome, you are connected to node orbstack.
+Running on Pod echo-bd94b7dcc-gdnhl.
+In namespace default.
+With IP address 192.168.194.18.
+Through HTTPS connection.
+```
+
+{{ site.base_gateway }} is now verifying the certificate of the upstream service and accepting the connection because
+the certificate is trusted.
+
+### Configure verification depth
+
+By default, {{ site.base_gateway }} verifies the certificate chain up to the root CA certificate with no depth limit.
+You can configure the verification depth by annotating the service with the `konghq.com/tls-verify-depth` annotation.
+
+For example, to limit the verification depth to 1 (i.e., only verify one intermediate certificate), you can annotate the
+service like this:
+
+```shell
+kubectl annotate service echo konghq.com/tls-verify-depth=1
+```
+
+The results should look like this.
+
+```text
+service/echo annotated
+```
+
+Now, when you issue the same request as before, you should still see a successful response.
+
+```shell
+curl -H Host:kong.example $PROXY_IP/echo
+```
+
+```text
+Welcome, you are connected to node orbstack.
+Running on Pod echo-bd94b7dcc-gdnhl.
+In namespace default.
+With IP address
+Through HTTPS connection.
+```
+
+You can also set the verification depth to 0 to not allow any intermediate certificates.
+
+```shell
+kubectl annotate --overwrite service echo konghq.com/tls-verify-depth=0
+```
+
+The results should look like this.
+
+```text
+service/echo annotated
+```
+
+Now, when you issue the same request as before, you should see an error similar to this.
+
+```shell
+curl -H Host:kong.example $PROXY_IP/echo
+```
+
+```text
+{
+  "message":"An invalid response was received from the upstream server",
+  "request_id":"e2b3182856c96c23d61e880d0a28012f"
+}
+```
```

Comment:
Hmm I couldn't get this so I started looking around.

I get an event:

```
37m (x25 over 38m)   Warning   KongConfigurationTranslationFailed   Secret/root-ca              invalid CA certificate: invalid PEM block
```

My cert looks ok though:
```
kubectl get secrets root-ca  -o jsonpath='{.data.cert}' | base64 -d  | openssl x509 -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            3f:ca:19:83:e8:9d:d7:24:dc:98:11:db:22:40:46:97:cf:9f:90:0d
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=root
        Validity
            Not Before: Dec 11 23:01:11 2024 GMT
            Not After : Dec 11 23:01:11 2025 GMT
        Subject: CN=root
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:c2:ae:23:d6:39:36:54:66:21:8e:31:04:ea:d8:
                    b6:0c:bc:fc:30:82:f9:1d:59:65:6e:0a:db:db:bb:
                    20:23:04:17:2b:d1:02:d2:d5:df:64:15:54:32:7d:
                    7e:27:0a:46:87:f0:84:25:5b:9d:31:3e:72:aa:47:
                    65:c9:59:52:12:5a:dd:97:36:eb:78:f9:6c:55:87:
                    85:35:e1:ae:8c:de:f6:b1:da:3a:44:26:67:60:28:
                    7f:da:92:25:e7:da:1e:c2:98:a3:57:c5:a1:51:61:
                    48:b8:cb:71:01:8e:2b:af:13:ad:b3:93:2b:a5:50:
                    22:79:c6:bb:f1:b3:d0:25:35:01:1c:c7:9c:f9:82:
                    29:5a:91:c9:93:64:a2:fe:77:6d:02:a8:7a:ec:96:
                    3d:0a:c8:a8:ad:b8:d1:e9:b8:90:7b:8e:74:24:26:
                    44:f8:63:a9:26:6d:ac:8e:ee:2f:0b:0f:37:1a:99:
                    e2:91:64:95:bc:a8:0d:71:93:ce:ee:c5:10:7e:5e:
                    d3:ff:5c:1b:dc:da:5b:3d:fe:37:88:02:3a:3f:08:
                    e8:42:d0:78:47:f8:90:66:65:c2:73:2f:eb:2f:8d:
                    d7:ad:fe:20:7d:46:aa:04:ef:33:a9:ee:a8:10:25:
                    ce:9f:c9:b5:d2:73:b0:53:7c:70:77:f3:8e:b6:6d:
                    5f:43
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Subject Key Identifier:
                51:0B:07:60:F2:70:F1:22:87:20:24:44:69:CB:6B:92:E0:88:A6:9E
            X509v3 Authority Key Identifier:
                51:0B:07:60:F2:70:F1:22:87:20:24:44:69:CB:6B:92:E0:88:A6:9E
            X509v3 Basic Constraints: critical
                CA:TRUE
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        49:bf:d2:45:7f:3d:be:c7:b4:59:6c:19:3c:5a:f9:1d:ca:5c:
        b0:77:9d:25:19:66:1a:4f:07:f9:88:30:2e:6d:dd:ce:96:55:
        9f:72:d6:b2:98:a8:a8:d4:ec:b1:bc:5d:90:60:00:e6:d0:18:
        63:46:ec:ec:1a:86:e8:6e:b4:30:fa:e3:a1:51:a7:ec:e3:1a:
        d7:34:0f:d6:e2:a7:52:89:14:0c:3b:4a:fd:a5:29:1c:e4:fd:
        d1:0f:fb:ed:a9:db:03:62:93:12:04:15:7b:44:bc:d4:7d:a3:
        53:fb:c7:42:9a:40:23:f8:b0:15:87:29:24:7b:ab:b8:a5:76:
        a7:ae:c9:f7:51:39:54:b7:d8:bb:87:d3:32:f9:59:02:5d:8f:
        0f:4b:d2:4c:71:c6:d6:21:a7:1b:cf:6a:2e:31:95:0e:94:74:
        be:54:94:de:48:a6:55:6d:6c:06:ff:ff:af:37:fa:b5:88:d8:
        24:78:05:07:56:16:e7:14:72:0b:ea:df:f6:4f:89:6c:9c:00:
        96:9b:24:23:61:2a:7c:e2:05:f4:34:41:04:67:97:fc:e8:95:
        88:7d:bd:8f:c3:09:e5:cb:30:e7:ad:73:16:c0:74:fc:4f:94:
        78:0d:7a:2c:03:49:37:36:64:b0:51:8f:3d:21:66:a6:20:0c:
        bd:10:51:bf
-----BEGIN CERTIFICATE-----
MIIC/zCCAeegAwIBAgIUP8oZg+id1yTcmBHbIkBGl8+fkA0wDQYJKoZIhvcNAQEL
BQAwDzENMAsGA1UEAwwEcm9vdDAeFw0yNDEyMTEyMzAxMTFaFw0yNTEyMTEyMzAx
MTFaMA8xDTALBgNVBAMMBHJvb3QwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEK
AoIBAQDCriPWOTZUZiGOMQTq2LYMvPwwgvkdWWVuCtvbuyAjBBcr0QLS1d9kFVQy
fX4nCkaH8IQlW50xPnKqR2XJWVISWt2XNut4+WxVh4U14a6M3vax2jpEJmdgKH/a
kiXn2h7CmKNXxaFRYUi4y3EBjiuvE62zkyulUCJ5xrvxs9AlNQEcx5z5gilakcmT
ZKL+d20CqHrslj0KyKituNHpuJB7jnQkJkT4Y6kmbayO7i8LDzcameKRZJW8qA1x
k87uxRB+XtP/XBvc2ls9/jeIAjo/COhC0HhH+JBmZcJzL+svjdet/iB9RqoE7zOp
7qgQJc6fybXSc7BTfHB38462bV9DAgMBAAGjUzBRMB0GA1UdDgQWBBRRCwdg8nDx
IocgJERpy2uS4IimnjAfBgNVHSMEGDAWgBRRCwdg8nDxIocgJERpy2uS4IimnjAP
BgNVHRMBAf8EBTADAQH/MA0GCSqGSIb3DQEBCwUAA4IBAQBJv9JFfz2+x7RZbBk8
Wvkdylywd50lGWYaTwf5iDAubd3OllWfctaymKio1OyxvF2QYADm0BhjRuzsGobo
brQw+uOhUafs4xrXNA/W4qdSiRQMO0r9pSkc5P3RD/vtqdsDYpMSBBV7RLzUfaNT
+8dCmkAj+LAVhykke6u4pXanrsn3UTlUt9i7h9My+VkCXY8PS9JMccbWIacbz2ou
MZUOlHS+VJTeSKZVbWwG//+vN/q1iNgkeAUHVhbnFHIL6t/2T4lsnACWmyQjYSp8
4gX0NEEEZ5f86JWIfb2PwwnlyzDnrXMWwHT8T5R4DXosA0k3NmSwUY89IWamIAy9
EFG/
-----END CERTIFICATE-----
```

I'll check again I must have done something wrong...

Reply:
Ok so I think this was the same problem as before. Just needed to wait some time with no traffic before it worked.

Reply:
I believe this is caused by `upstream_keepalive_idle_timeout` (which is 60 - seconds - by default). Added a note about that in 25efd421af37161548dea6914a936f09ab52dfe3.